runtime/convertValues.go

```diff
@@ -256,36 +266,118 @@ func exportCompositeValue(
 	error,
 ) {
-	pairs := make([]cadence.KeyValuePair, v.Count())
+	dynamicType, ok := v.DynamicType(inter, interpreter.SeenReferences{}).(interpreter.CompositeDynamicType)
+	if !ok {
+		return nil, fmt.Errorf(
+			"unexportable composite value: %s", dynamicType.StaticType,
+		)
+	}
+	staticType := dynamicType.StaticType.(*sema.CompositeType)
+	// TODO: consider making the results map "global", by moving it up to exportValueWithInterpreter
+	t := exportCompositeType(staticType, map[sema.TypeID]cadence.Type{})
 
-	for i, keyValue := range v.Keys().Elements() {
+	// NOTE: use the exported type's fields to ensure fields in type
+	// and value are in sync
 
-		// NOTE: use `Get` instead of accessing `Entries`,
-		// so that the potentially deferred values are loaded from storage
+	fieldNames := t.CompositeFields()
+	fields := make([]cadence.Value, len(fieldNames))
 
-		value := v.Get(inter, interpreter.ReturnEmptyLocationRange, keyValue).(*interpreter.SomeValue).Value
+	for i, field := range fieldNames {
+		fieldName := field.Identifier
+
+		fieldValue := v.Fields[fieldName]
+		if fieldValue == nil && v.ComputedFields != nil {
+			if computedField, ok := v.ComputedFields[fieldName]; ok {
+				// TODO: provide proper location range
+				fieldValue = computedField(inter, interpreter.ReturnEmptyLocationRange)
+			}
+		}
 
-		convertedKey, err := exportValueWithInterpreter(keyValue, inter, seenReferences)
+		exportedFieldValue, err := exportValueWithInterpreter(fieldValue, inter, seenReferences)
 		if err != nil {
-			return cadence.Dictionary{}, err
+			return nil, err
 		}
-		convertedValue, err := exportValueWithInterpreter(value, inter, seenReferences)
+		fields[i] = exportedFieldValue
+	}
+
+	// NOTE: when modifying the cases below,
+	// also update the error message below!
+
+	switch staticType.Kind {
+	case common.CompositeKindStructure:
+		return cadence.NewStruct(fields).WithType(t.(*cadence.StructType)), nil
+	case common.CompositeKindResource:
+		return cadence.NewResource(fields).WithType(t.(*cadence.ResourceType)), nil
+	case common.CompositeKindEvent:
+		return cadence.NewEvent(fields).WithType(t.(*cadence.EventType)), nil
+	case common.CompositeKindContract:
+		return cadence.NewContract(fields).WithType(t.(*cadence.ContractType)), nil
+	case common.CompositeKindEnum:
+		return cadence.NewEnum(fields).WithType(t.(*cadence.EnumType)), nil
+	}
+
+	return nil, fmt.Errorf(
+		"invalid composite kind `%s`, must be %s",
+		staticType.Kind,
+		common.EnumerateWords(
+			[]string{
+				common.CompositeKindStructure.Name(),
+				common.CompositeKindResource.Name(),
+				common.CompositeKindEvent.Name(),
+				common.CompositeKindContract.Name(),
+				common.CompositeKindEnum.Name(),
+			},
+			"or",
+		),
+	)
+}
+
+func exportDictionaryValue(
+	v *interpreter.DictionaryValue,
+	inter *interpreter.Interpreter,
+	seenReferences seenReferences,
+) (
+	cadence.Dictionary,
+	error,
+) {
+	pairs := make([]cadence.KeyValuePair, 0, v.Count())
```

Comment:
Same as above, dictionaries do not have an `Elements()` function anymore, replace it with iteration